src/primitives/cl_dl_public_setup.rs

```diff
@@ -0,0 +1,580 @@
+//! Implementation verifiable encryption of discrete logarithms in the Castagnos
+//! and Laguillaumie cryptosystem where the CL group has a public setup.
+//!
+//!
+//! The advantage of the CL cryptosystem is that you can generate a CL-group of
+//! unknown order by where the discrete logarithm is **easy** for a subgroup of
+//! an order of your choosing. For example, this allows you to generate a group
+//! where you can verifiably encrypt the private key of a Bitcoin public key.
+//!
+//! 1. CL-groups were defined in https://eprint.iacr.org/2015/047.pdf, along
+//!    with the original CPA-secure encryption scheme.
+//! 2. The particular group generation algorithm, where you pass in the order of
+//!    the DL-easy subgroup was shown in https://eprint.iacr.org/2019/503.pdf.
+//! 3. The proof of equivalence between the discrete logarithm of a DL group
+//!    element and the plaintext of a CL ciphertext in a CL-group with a public
+//!    setup is given in https://eprint.iacr.org/2020/084.pdf (see section 5.2)
+
+use super::ErrorReason;
+use crate::bn_to_gen;
+use crate::curv::arithmetic::traits::Modulo;
+use crate::curv::cryptographic_primitives::hashing::traits::Hash;
+use crate::isprime;
+use crate::pari_init;
+use crate::primitives::is_prime;
+use crate::primitives::numerical_log;
+use crate::primitives::prng;
+use crate::BinaryQF;
+use curv::arithmetic::traits::Samplable;
+use curv::cryptographic_primitives::hashing::hash_sha256::HSha256;
+use curv::elliptic::curves::traits::{ECPoint, ECScalar};
+use curv::BigInt;
+use curv::{FE, GE};
+use std::os::raw::c_int;
+
+const SECURITY_PARAMETER: usize = 128;
+const C: usize = 10;
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct CLGroup {
+    pub delta_k: BigInt,
+    pub delta_q: BigInt,
+    pub gq: BinaryQF,
+    pub stilde: BigInt,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub struct Ciphertext {
+    pub c1: BinaryQF,
+    pub c2: BinaryQF,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct TTriplets {
+    pub t1: BinaryQF,
+    pub t2: BinaryQF,
+    pub T: GE,
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize)]
+pub struct U1U2 {
+    pub u1: BigInt,
+    pub u2: BigInt,
+}
+
+#[derive(Debug, Clone)]
+pub struct ProofError;
+
+impl CLGroup {
+    pub fn new_from_setup(lam: &usize, seed: &BigInt) -> Self {
+        let q = &FE::q();
+        unsafe { pari_init(100000000, 2) };
+        let mu = q.bit_length();
+        debug_assert!(lam > &(mu + 2));
```

Comment:
Note: I think I also changed these from `assert` to `debug_assert`. Please let me know if that wasn't what was intended but this would be a weird place to panic.

Reply:
If Lam is provided as input this should not be a debugging assertion. Probably better to return Result with error. At current state of the library I don't mind keeping assert/debug_assert

Reply:
Right actually this makes more sense as an `assert` I think it was actually the other one in this function that was less clear. I'll change it to an assert.